drivers/tee/optee/optee.c

```diff
@@ -88,10 +92,38 @@ struct optee_driver_data {
 	unsigned long sec_caps;
 };
 
+#ifdef CONFIG_MMU
+/*
+ * Use arch_page_phys_get instead of z_mem_phys_addr when MMU
+ * is enabled.
+ * According to the description of z_mem_phys_addr it uses
+ * Z_MEM_VM_OFFSET to make a conversion from virtual to physical
+ * address. And these macros are intended for assembly, linker code,
+ * and static initializers.
+ * So when external libraries, such as PICOLIBC or NEWLIB_LIBC
+ * uses internal mappings for dynamic memory allocation z_mem_phys_addr
+ * may not work properly and return wrong physical address.
+ */
+static uintptr_t optee_mem_phys_addr(void *virt)
+{
+	uintptr_t phys;
+	uint32_t offt = (uintptr_t)virt % CONFIG_MMU_PAGE_SIZE;
```

Comment:
It's a pity that zephyr does not provide `PAGE_MASK`. I just hope that compiller will be smart enough to replace division with bitwise AND.

Reply:
fixed